docs/design/sparse-v2.md

```diff
@@ -0,0 +1,225 @@
+
+All Sparse Patterns v2 will come equipped with a default no-op mapping.
+
+```rust
+vec![WorkingCopyMap {
+    src_path: RepoPathBuf::root(),
+    dst_path: RepoPathBuf::root(),
+}]
+```
+
+`SparsePatterns` will provide an interface to map client paths into repo paths,
+and vice versa. The WorkingCopy trait will apply this mapping to all snapshot
+and checkout operations, and and jj commands which accept relative paths will
+need to be updated to perform client path -> repo path translations as needed.
+It's not clear at this time _which_ commands will need changing, as some are
+more likely to refer to repo paths rather than client paths.
+
+TODO: Expand this section.
+
+In particular, the path rules for sparse patterns for _always_ be repo paths,
+not client paths. Thus, if the client wants to track "foo" and rename it to
+"subdir/bar", they must `jj sparse set --add foo` and
+`jj client-map set --from foo --to bar`. In other words, the mapping operation can
+be thought of as always _after_ the sparse operation.
+
+New commands will enable editing of the WorkingCopyMap:
+
+- `jj client-map list` will print all mapping pairs.
+- `jj client-map set --from foo --to bar` will add a new mapping to the end of the list.
+- `jj client-map remove --from foo` will remove a specific mapping rule.
+- `jj client-map edit` will pull up a text editor for manual editing.
+
+Like sparse paths, mapping rules are defined to apply in _order_ and on any
+save operation will be modified to a minimal canonical form. Thus,
+`jj client-map set --from "" --to ""` will always completely wipe the map.
+The first matching rule in reverse list order determines how a particular
+folder/file should be named in the client.
+
+For simplicity, we require that all client-maps are _invertible_, that is:
+
+- For any repo path R, there is at most one associated client path C
+- Having no associated client path is fine (that part of the repo will never be materialized)
+- For any client path C, there is at most one associated repo path R
+- Having no associated repo path is fine (all files in such paths will be ignored)
+
+To this end, recursive mappings are considers _destructive_. That is, if "foo"
+is renamed to "bar", but "bar" already exists in the repo, then the mapping rule
+hides the original "bar" directory, making it uneditable in the client.
+
+## Versioning and Storage
+
+Updating the active Sparse Pattern for a particular working copy will now
+take place in two separate steps, one which updates the Op Store, and
+another which updates the actual working copy.
+
+This gives the user the ability to update the active sparse patterns whilst
+not interacting with the local working copy, which is useful for custom
+integrations which may not be _able_ to check out particular sparse patterns due
+to problems with the backend (encoding, permission errors, etc.). A bad
+`jj sparse set --add oops` command can thus be undone, even via `jj op undo`
+if desired.
```

Comment:
This kinda implies that `jj sparse` will always call `jj sync`. It also leaves DDOS'ing abilities in the hand of clients, I assume this already dealt with in the Google internal backend but may hinder the development of other backends.
 
Because as I've read it, I've felt that a `jj sparse set 'include:path/to/file'` updates the view immediately w/o considering that the path isn't available yet.

Reply:
> This kinda implies that `jj sparse` will always call `jj sync`.
 
Why do you think so? The scenario here is if you do `jj sparse set --add <secret directory I don't have access to>`, then you can undo do that with `jj --ignore-working-copy undo`.

Reply:
> > This kinda implies that `jj sparse` will always call `jj sync`.
> 
> Why do you think so?

Because there's no explanation _when_ the view change applies. My immediate thought was to only apply it after `jj sync` (when the client updates) without losing pending work in the current working copy. 

The way it is currently worded, it could also imply that we directly update the view and lose all the work which is no longer within the view, which is kinda bad. 

> The scenario here is if you do `jj sparse set --add <secret directory I don't have access to>`, then you can undo do that with `jj --ignore-working-copy undo`

Ack, that was clear from the beginning. Why the `jj --ignore-working-copy undo` though? Shouldn't the plain `jj undo` just work or does it leave the working-copy in a uncertain state?

Reply:
> The way it is currently worded, it could also imply that we directly update the view

That was my assumption.

> and lose all the work which is no longer within the view, which is kinda bad.

Since we snapshot the working copy at the start of every command, I don't think we'll lose any work. Any changes in the working copy will go into the new snapshot before we update the working copy to no longer have those files. I think it will work the same in other workspaces too (no data loss).

> Why the `jj --ignore-working-copy undo` though?

It wouldn't typically be needed, but it would be needed in the case where you have files in directories you don't have access to, because snapshotting those directories probably fails (that's the reason you're now trying to undo).

Reply:
> Since we snapshot the working copy at the start of every command, I don't think we'll lose any work. Any changes in the working copy will go into the new snapshot before we update the working copy to no longer have those files. I think it will work the same in other workspaces too (no data loss).

What happens exactly if I have pending changes in `<secret>/dir/file.rs` and then change the view to exclude `<secret>/`? It will be in the snapshot but no longer visible, so something a user may perceive as "data loss". 

> > Why the `jj --ignore-working-copy undo` though?
> 
> It wouldn't typically be needed, but it would be needed in the case where you have files in directories you don't have access to, because snapshotting those directories probably fails (that's the reason you're now trying to undo).

I dislike that we have to ignore the working copy for a "clean" undo, it feels very inconsistent. Maybe having preciser ACL metadata let's to avoid it?

Reply:
> What happens exactly if I have pending changes in `<secret>/dir/file.rs` and then change the view to exclude `<secret>/`? It will be in the snapshot but no longer visible, so something a user may perceive as "data loss".

If you have changes in the secret directory, then they would not be snapshotted and would remain in the working copy when you exclude it again. But you would probably notice as soon as you added the secret directory with `jj sparse set --add <secret>` that it failed, so it seems unlikely that users start adding content after that.

> I dislike that we have to ignore the working copy for a "clean" undo, it feels very inconsistent. Maybe having preciser ACL metadata let's to avoid it?

It may actually not be needed. Since updating the working copy to match the repo's view of the sparse patterns failed, the working copy would still have the old patterns, so maybe snapshotting would just work.

Reply:
> If you have changes in the secret directory, then they would not be snapshotted and would remain in the working copy when you exclude it again.

So the changes will stick around until I can get access again? Wouldn't it be easier to just add them to the previous commit and do rename tracking and co, until you can unhide them again?  

> But you would probably notice as soon as you added the secret directory with `jj sparse set --add <secret>` that it failed, so it seems unlikely that users start adding content after that.

Sorry, I do not get that. My usecase describes the opposite workflow, when you first have access and have changes in the directory and then lose access.  

> It may actually not be needed. Since updating the working copy to match the repo's view of the sparse patterns failed, the working copy would still have the old patterns, so maybe snapshotting would just work.

That seems great.

Reply:
Sorry, what is `jj sync`?  I think I am missing some context here.

I've expanded the section on 'Versioning and Storage' to hopefully be clearer about the intended order of operations.

Reply:
Okay I am more lost then - why would `jj sparse` or `jj map` invoke that?  I'm not sure what part of the text I wrote suggests that, I definitely didn't intend to suggest it.

By 'update the Op Store' I only meant 'write a new Op Heads + Operation pointing to an update View object'

Reply:
For _me_ it wasn't clear when the view updates, so quick rationalization led to this implication. Which is wrong and something you didn't write. 

From the conversation with Martin above:
> > > This kinda implies that `jj sparse` will always call `jj sync`.
> > 
> > 
> > Why do you think so?
> 
> Because there's no explanation _when_ the view change applies. My immediate thought was to only apply it after `jj sync` (when the client updates) without losing pending work in the current working copy.

> I'm not sure what part of the text I wrote suggests that, I definitely didn't intend to suggest it.

Yes, that's clearly my fault, you did nothing wrong here.

Reply:
Resolving this, as I don't want to discuss my missunderstanding again.